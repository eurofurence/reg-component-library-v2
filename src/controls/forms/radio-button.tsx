/** @jsxImportSource @emotion/react */

import {
	ChangeEventHandler,
	Fragment,
	ReactNode,
	createContext,
	forwardRef,
	useContext,
} from "react"
import { css } from "@emotion/react"
import styled from "@emotion/styled"

import type { DeepReadonly } from "ts-essentials"

import FieldSet from "./field-set"
import { WithFormLabelProps, withFormLabel } from "./form-label"

const NameContext = createContext<{
	readonly name: string
	readonly invalid: boolean
}>({ name: "unknown", invalid: false })

const Input = styled.input<{ readonly invalid: boolean }>`
	// TODO: Probably scale padding and border along.

	width: 1.25em;
	height: 1.25em;
	border: 2px solid var(--color-grays-300);
	border-radius: 0.625em;

	&:checked {
		background-color: var(--color-semantic-info);
		border-color: var(--color-semantic-info);
		background-clip: content-box;
		padding: 3px;
	}

	${({ invalid = false }) =>
		!invalid
			? css``
			: css`
		border-color: var(--color-semantic-error);
	`}

	&:invalid {
		border-color: var(--color-semantic-error);
	}
`

interface PlainRadioItemProps {
	readonly value: string
	readonly checked?: boolean
	readonly defaultChecked?: boolean
	readonly onChange?: ChangeEventHandler<HTMLInputElement>
	readonly readOnly?: boolean
}

export type RadioItemProps = WithFormLabelProps<PlainRadioItemProps>


export const RadioItem = withFormLabel<HTMLInputElement, PlainRadioItemProps>(
	forwardRef<HTMLInputElement, PlainRadioItemProps>((props, ref) => {
		const { name, invalid } = useContext(NameContext)

		return (
			<Input {...props} name={name} type="radio" invalid={invalid} ref={ref} />
		)
	})
)

export interface RadioGroupProps {
	readonly name: string
	readonly invalid?: boolean
	readonly children: DeepReadonly<ReactNode>
}

export const RadioGroup = ({
	name,
	invalid = false,
	children,
}: RadioGroupProps) => (
	<Fragment>
		<NameContext.Provider value={{ name, invalid }}>
			{children}
		</NameContext.Provider>
	</Fragment>
)

export interface RadioSetProps {
	readonly name: string
	readonly gridSpan?: number
	readonly legend: string
	readonly error?: string
	readonly children: DeepReadonly<ReactNode>
}

export const RadioSet = ({
	name,
	gridSpan,
	legend,
	error,
	children,
}: RadioSetProps) => (
	<FieldSet legend={legend} gridSpan={gridSpan} error={error}>
		<RadioGroup name={name} invalid={Boolean(error)}>
			{children}
		</RadioGroup>
	</FieldSet>
)

RadioGroup.Item = RadioItem
RadioSet.Item = RadioItem

export default RadioGroup
