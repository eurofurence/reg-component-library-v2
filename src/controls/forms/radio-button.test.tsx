/** @jsxImportSource @emotion/react */

import { render } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import * as fc from "fast-check"
import renderer from "react-test-renderer"

import { RadioGroup, RadioItem, RadioSet } from "./radio-button"

describe("RadioItem", () => {
	it("renders correctly", () => {
		fc.assert(
			fc.property(
				fc.option(fc.string(), { nil: undefined }),
				fc.boolean(),
				(label, checked) => {
					const tree = renderer
						.create(
							<RadioGroup name="popularity-contest">
								<RadioItem value="nick-wilde" label={label} checked={checked} />
							</RadioGroup>
						)
						.toJSON()

					expect(tree).toMatchSnapshot()
				}
			),
			{ numRuns: 20, seed: 2562346 }
		)
	})

	it("has a label if one is provided", () => {
		const { queryByLabelText } = render(
			<RadioGroup name="popularity-contest">
				<RadioItem value="nick-wilde" label="Nick Wilde" />
			</RadioGroup>
		)

		expect(queryByLabelText("Nick Wilde")).toBeTruthy()
	})

	it("does not have a label if none is provided", () => {
		const { container } = render(
			<RadioGroup name="popularity-contest">
				<RadioItem value="nick-wilde" />
			</RadioGroup>
		)

		expect(container.querySelector("label")).toBeFalsy()
	})

	describe("when controlled", () => {
		it("reflects the input value at all times (true)", async () => {
			const onChange = jest.fn()
			const { getByRole } = render(
				<RadioGroup name="popularity-contest">
					<RadioItem
						value="nick-wilde"
						label="Nick Wilde"
						checked={true}
						onChange={onChange}
					/>
				</RadioGroup>
			)

			const radio = getByRole("radio")

			expect(radio).toBeChecked()

			await userEvent.click(radio)

			expect(radio).toBeChecked()
		})

		it("reflects the input value at all times (false)", async () => {
			const onChange = jest.fn()
			const { getByRole } = render(
				<RadioGroup name="popularity-contest">
					<RadioItem
						value="nick-wilde"
						label="Nick Wilde"
						checked={false}
						onChange={onChange}
					/>
				</RadioGroup>
			)

			const radio = getByRole("radio")

			expect(radio).not.toBeChecked()

			await userEvent.click(radio)

			expect(radio).not.toBeChecked()
		})

		it("raises onChange events when clicked directly", async () => {
			const onChange = jest.fn()
			const { getByRole } = render(
				<RadioGroup name="popularity-contest">
					<RadioItem
						value="nick-wilde"
						label="Nick Wilde"
						checked={false}
						onChange={onChange}
					/>
				</RadioGroup>
			)

			const radio = getByRole("radio")

			await userEvent.click(radio)

			expect(onChange).toHaveBeenCalled()
		})

		it("raises onChange events when clicked indirectly", async () => {
			const onChange = jest.fn()
			const { getByText } = render(
				<RadioGroup name="popularity-contest">
					<RadioItem
						value="nick-wilde"
						label="Nick Wilde"
						checked={false}
						onChange={onChange}
					/>
				</RadioGroup>
			)

			const radio = getByText("Nick Wilde")

			await userEvent.click(radio)

			expect(onChange).toHaveBeenCalled()
		})
	})

	describe("when uncontrolled", () => {
		it("toggles when clicked directly", async () => {
			const { getByRole } = render(
				<RadioGroup name="popularity-contest">
					<RadioItem value="nick-wilde" label="Nick Wilde" />
				</RadioGroup>
			)

			const radio = getByRole("radio")

			expect(radio).not.toBeChecked()

			await userEvent.click(radio)

			expect(radio).toBeChecked()
		})

		it("toggles when clicked through label", async () => {
			const { getByRole, getByText } = render(
				<RadioGroup name="popularity-contest">
					<RadioItem value="nick-wilde" label="Nick Wilde" />
				</RadioGroup>
			)

			const label = getByText("Nick Wilde")
			const radio = getByRole("radio")

			expect(radio).not.toBeChecked()

			await userEvent.click(label)

			expect(radio).toBeChecked()
		})

		it("reflects the defaultChecked value (true) originally", () => {
			const { getByRole } = render(
				<RadioGroup name="popularity-contest">
					<RadioItem
						value="nick-wilde"
						label="Nick Wilde"
						defaultChecked={true}
					/>
				</RadioGroup>
			)

			expect(getByRole("radio")).toBeChecked()
		})

		it("reflects the defaultChecked value (false) originally", () => {
			const { getByRole } = render(
				<RadioGroup name="popularity-contest">
					<RadioItem
						value="nick-wilde"
						label="Nick Wilde"
						defaultChecked={false}
					/>
				</RadioGroup>
			)

			expect(getByRole("radio")).not.toBeChecked()
		})

		it("raises onChange events when toggled directly", async () => {
			const onChange = jest.fn()
			const { getByRole } = render(
				<RadioGroup name="popularity-contest">
					<RadioItem
						value="nick-wilde"
						label="Nick Wilde"
						onChange={onChange}
					/>
				</RadioGroup>
			)

			const radio = getByRole("radio")

			await userEvent.click(radio)

			expect(onChange).toHaveBeenCalled()
		})

		it("raises onChange events when toggled indirectly", async () => {
			const onChange = jest.fn()
			const { getByText } = render(
				<RadioGroup name="popularity-contest">
					<RadioItem
						value="nick-wilde"
						label="Nick Wilde"
						onChange={onChange}
					/>
				</RadioGroup>
			)

			const radio = getByText("Nick Wilde")

			await userEvent.click(radio)

			expect(onChange).toHaveBeenCalled()
		})
	})
})

describe("RadioGroup", () => {
	it("groups radio buttons together", async () => {
		const onChange = jest.fn()
		const { getByLabelText } = render(
			<RadioGroup name="popularity-contest">
				<RadioItem value="robin-hood" label="Robin Hood" onChange={onChange} />
				<RadioItem value="nick-wilde" label="Nick Wilde" onChange={onChange} />
				<RadioItem
					value="fox-mccloud"
					label="Fox McCloud"
					onChange={onChange}
				/>
			</RadioGroup>
		)

		const robinHood = getByLabelText("Robin Hood")
		const nickWilde = getByLabelText("Nick Wilde")

		await userEvent.click(robinHood)

		expect(robinHood).toBeChecked()

		await userEvent.click(nickWilde)

		expect(robinHood).not.toBeChecked()
	})
})

describe("RadioSet", () => {
	it("groups radio buttons together", async () => {
		const onChange = jest.fn()
		const { getByLabelText } = render(
			<RadioSet name="popularity-contest" legend="Who is best fox?">
				<RadioItem value="robin-hood" label="Robin Hood" onChange={onChange} />
				<RadioItem value="nick-wilde" label="Nick Wilde" onChange={onChange} />
				<RadioItem
					value="fox-mccloud"
					label="Fox McCloud"
					onChange={onChange}
				/>
			</RadioSet>
		)

		const robinHood = getByLabelText("Robin Hood")
		const nickWilde = getByLabelText("Nick Wilde")

		await userEvent.click(robinHood)

		expect(robinHood).toBeChecked()

		await userEvent.click(nickWilde)

		expect(robinHood).not.toBeChecked()
	})

	it("adds a legend", () => {
		const onChange = jest.fn()
		const { queryByText } = render(
			<RadioSet name="popularity-contest" legend="Who is best fox?">
				<RadioItem value="robin-hood" label="Robin Hood" onChange={onChange} />
				<RadioItem value="nick-wilde" label="Nick Wilde" onChange={onChange} />
				<RadioItem
					value="fox-mccloud"
					label="Fox McCloud"
					onChange={onChange}
				/>
			</RadioSet>
		)

		expect(queryByText("Who is best fox?")).toBeTruthy()
	})
})
