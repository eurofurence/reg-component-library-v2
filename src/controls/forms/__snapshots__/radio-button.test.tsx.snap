// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioItem renders correctly 1`] = `null`;

exports[`RadioItem renders correctly 2`] = `null`;

exports[`RadioItem renders correctly 3`] = `null`;

exports[`RadioItem renders correctly 4`] = `null`;

exports[`RadioItem renders correctly 5`] = `null`;

exports[`RadioItem renders correctly 6`] = `null`;

exports[`RadioItem renders correctly 7`] = `null`;

exports[`RadioItem renders correctly 8`] = `null`;

exports[`RadioItem renders correctly 9`] = `null`;

exports[`RadioItem renders correctly 10`] = `null`;

exports[`RadioItem renders correctly 11`] = `null`;

exports[`RadioItem renders correctly 12`] = `null`;

exports[`RadioItem renders correctly 13`] = `null`;

exports[`RadioItem renders correctly 14`] = `null`;

exports[`RadioItem renders correctly 15`] = `null`;

exports[`RadioItem renders correctly 16`] = `null`;

exports[`RadioItem renders correctly 17`] = `null`;

exports[`RadioItem renders correctly 18`] = `null`;

exports[`RadioItem renders correctly 19`] = `null`;

exports[`RadioItem renders correctly 20`] = `null`;
